source/server/listener_impl.cc

```diff
@@ -360,10 +367,17 @@ ListenerImpl::ListenerImpl(ListenerImpl& origin,
           origin.listener_factory_context_->listener_factory_context_base_, this, *this)),
       filter_chain_manager_(address_, origin.listener_factory_context_->parentFactoryContext(),
                             initManager(), origin.filter_chain_manager_),
-      local_init_watcher_(fmt::format("Listener-local-init-watcher {}", name), [this] {
-        ASSERT(workers_started_);
-        parent_.inPlaceFilterChainUpdate(*this);
-      }) {
+      local_init_watcher_(fmt::format("Listener-local-init-watcher {}", name),
+                          [this] {
+                            ASSERT(workers_started_);
+                            parent_.inPlaceFilterChainUpdate(*this);
+                          }),
+      transport_factory_context_(parent_.server_.admin(), parent_.server_.sslContextManager(),
```

Comment:
Sorry, this breaks intelligent listener update.
Previously the transport_factory_context was on the stack. 
Now the transport_factory_context instance needs to be extended because this instance is referred by tls transport socket `ContextConfigImpl`'s callback.

Suppose you have listener config with filter chain X, and new listener config with filter chain X+Y, 
the listener update will keep the previous X (and its TLS transport socket!) valid,
but the old listener is destroyed so the old transport_factory_context_ is destroyed as well. The filter chain X ends up with a dangling reference to the old transport_factory_context_.

I suggest defining ListenerImpl::transport_factory_context_ as shared_ptr and share the transport_factory_context_ among generations of ListenerImpl.
See the shared member `connection_balancer_(origin.connection_balancer_)`.

Reply:
Thanks for the feedback! I guess I didn't quite understand how the listener config update is done at first. I added now a patch to use a shared pointer as you suggested. Could you see if it resolves the issue?

Reply:
Thanks! This part LGTM. I will use a new thread as follow up

Reply:
Could you add an integration test for this? My experience is that any combination of 2 xDS is very like to introduce issues. I don't fully sure what's the lifetime of the transport socket in `ContextConfigImpl::setSecretUpdateCallback`. Neither downstream nor upstream. It would be good to have an integration test in test/integration/sds_dynamic_integration_test.cc

Reply:
Since you are the original author of the private key provider :)

Is there an example of PrivateKeyMethodProviderInstanceFactory for me to learn? Thanks!

Reply:
Ok, I'll add an integration test. So far I've only been trying this out with modified go-control-plane examples. I'll try to experiment with various xDS combinations -- do you have any combinations in mind which would be most likely to expose lifetime issues?

I have a few internal private key providers in use, but so far the only example in the Envoy tree is the test private key provider at https://github.com/envoyproxy/envoy/blob/main/test/extensions/transport_sockets/tls/test_private_key_method_provider.cc. However, I'm planning to introduce an AVX-512 private key provider "soon", see https://github.com/envoyproxy/envoy/issues/15871.